src/minecraft/client.rs

```diff
@@ -63,166 +81,172 @@ impl Client {
     ///
-    /// tx.try_send(()); // Stop the packet read loop.
+    /// block_on(client.send_chat_message("Hello!")); // Send the chat message. Assumes success of connection
     /// ```
-    pub async fn start_loop(client: Arc<Mutex<Self>>, mut receiver: Receiver<()>) {
+    pub async fn send_chat_message(&self, msg: &str) {
+        let mut runner = self.runner.lock().await;
+        runner.send_chat_message(msg).await.unwrap();
+    }
+
+    /// Register a packet listener to intercept packets received by the server
+    pub async fn on_receive_packet(&self, cb: PacketListener) {
+        self.packet_listeners.lock().await.push(cb);
+    }
+
+    /// Attempts to receive a packet from the ClientRunner and calls the packet listeners
+    async fn receive_packet(packet_listeners: &mut Arc<Mutex<Vec<PacketListener>>>, recv: &mut mpsc::Receiver<Packet>) {
+        if let Some(packet) = recv.recv().await {
+            let pl_lock = packet_listeners.lock().await;
+            pl_lock.iter().for_each(|f| {
+                f(&packet);
+            })
+        }
+    }
+}
+
+impl ClientRunner {
+    /// Starts the client loop and listens for incoming server packets
+    async fn start_loop(client: Arc<Mutex<Self>>, mut receiver: Receiver<()>) {
         let client_arc = client.clone();
         loop {
             if let Err(err) = receiver.try_recv() {
                 if err == TryRecvError::Closed {
+                    println!("[Client] Closed receiver connection");
                     break;
                 }
-                let client_lock = &mut client_arc.lock().await;
+                let mut client_lock = client_arc.lock().await;
                 if !client_lock.connected {
+                    println!("[Client] Disconnected from server");
                     break;
                 }
                 let packet_read: Result<super::Packet>;
                 {
+                    // Attempt to read a client packet. Only attempt for 5 ms to allow others to take the lock
                     let delayer = client_lock.read_packet();
-                    pin_mut!(delayer);
                     if let Ok(packet) =
-                        tokio::time::timeout(std::time::Duration::from_millis(1), &mut delayer)
+                        tokio::time::timeout(std::time::Duration::from_millis(5), delayer)
                             .await
                     {
                         packet_read = packet;
                     } else {
                         continue;
                     }
                 }
-                if let Ok(packet) = packet_read {
-                    client_lock.handle_packet(packet).await;
-                };
+                // Process a packet if deserialized correctly
+                match packet_read {
+                    Ok(packet) => client_lock.handle_packet(packet).await,
+                    Err(e) => {
+                        println!("[Client] {e}")
```

Comment:
It would be interesting to use something like [`env_logger`](https://docs.rs/env_logger/latest/env_logger/) for better logs, but manual/println is fine for now.